### PR TITLE
feat(desktop): interactive to-do list (built-in widget + agent tool)

### PR DESCRIPTION
**Pull request by kyh**
## What & why

Exploring the todo/calendar features from [get-growth.app](https://www.get-growth.app/). This PR lands the **first slice: a real interactive to-do list**.

Today the desktop has a *static* "To Do" dashboard card (hardcoded checkboxes in `seed-widgets.ts`) and a read-only Google Calendar feed (`Up Next` / `Meeting Prep`). This adds a genuine, persistent, agent-aware to-do feature — the foundation a unified agenda/calendar view can build on next.

It's deliberately **distinct from the existing scheduled "tasks" system**: tasks are cron jobs the agent runs on its own; todos are the user's own checklist (priority + due dates).

## Approach

Mirrors the `TaskManager` pattern end to end, so it slots into the existing architecture rather than inventing a new one:

- **`shared/todo.ts`** — `Todo` schema, method params, and pure `sortTodos` / `formatDue` / `isOverdue` helpers (shared by the panel and the agent tool).
- **`main/todos/todo-manager.ts`** — versioned `JsonStore` at `~/.inteligir/todos.json`, CRUD, a change-notifier push seam, lazy singleton + reset (wired into `teardownResources`).
- **IPC** — `createTodo` / `listTodos` / `updateTodo` / `toggleTodo` / `deleteTodo` / `clearCompletedTodos` + an `onTodosUpdated` push channel. Handlers in `main/index.ts`; broadcast wired in `app-machine.ts`. (Bridge auto-derives from the registry.)
- **Renderer** — `todo-store` (Zustand, live via the push channel with optimistic updates) and a `TodoPanel` built-in widget (add form with priority pills + due date, overdue highlighting, clear-completed). `"todos"` added to the builtin registry, so it appears in the dock automatically.
- **Agent** — `TodosPort` on `AgentPorts` + a `manage_todos` tool (list/create/update/toggle/delete/clear_completed), and each turn is primed with the open list so "remind me to…" / "what's on my list?" work by voice.

## Tests / gates

- New `todo-manager` and `todos-extension` suites; existing `AgentPorts` mocks updated for the new port.
- `typecheck` ✓ · `lint` ✓ · `format` ✓ · `knip` ✓ · `build` ✓ · **517 tests** ✓

## Notes / follow-ups

- The static `TODO_WIDGET` demo card on the dashboard grid is left untouched for now, so there are briefly two "To Do" surfaces (the demo card + the real dock panel). Folding the card into the real store is a natural next step.
- Natural next slices from the planning discussion: a **unified agenda view** (merge these todos with Google Calendar events — get-growth's signature feature) and a **calendar grid**.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01RPC3YHCfsqCPrG9fR7sARG

---
_Generated by [Claude Code](https://claude.ai/code/session_01RPC3YHCfsqCPrG9fR7sARG)_

---

> [!NOTE]
> **Medium Risk**
> Two-way external sync and mid-flight reconciliation can affect user task data; mitigated by pagination aborts, tombstones, and applySync guards. Dashboard seed layout changes affect first-run UX.
> 
> **Overview**
> Introduces a **persistent user to-do list** (separate from scheduled agent tasks): `TodoManager` on versioned disk storage, full IPC + `onTodosUpdated` push, a **To Do** built-in panel with sync, and a **`manage_todos`** agent tool that primes each turn with open items.
> 
> Adds a **unified Agenda** built-in that merges paginated Google Calendar events with due-dated open todos via shared `buildAgenda` logic, with in-agenda todo completion wired to the same store.
> 
> Implements **two-way Google Tasks sync** through the executor (`planSync`, tombstones, single-flight, first-sync title dedup with done-state guards) and registers **google_tasks** in the connector catalog.
> 
> **Dashboard seeding** drops the static Meeting Prep / Up Next / static To Do json-ui cards and pins **Agenda** + **To Do** built-ins alongside the remaining json-ui widgets (`date`, `people`, `today`, `weather`). Logout/teardown resets todo state and suppresses stale broadcasts.
> 
> Broad **test coverage** for agenda merge, sync planning, manager CRUD/sync edge cases, and the agent extension.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9aaa6238cb34aff7e75e7a59fd92ed909aae92a9. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

---
## Summary by cubic
Adds a persistent to‑do list, a unified Agenda that merges Google Calendar events with due‑dated to‑dos, and two‑way Google Tasks sync with pagination, tombstones, and first‑sync title dedup (1:1 by title; matches only when done-state aligns). Replaces the static cards and aligns with Linear 2gx4hg (todo/calendar integration).

- **New Features**
  - Storage: versioned `~/.inteligir/todos.json` via `TodoManager` with CRUD, push notifications, atomic `applySync`; v2 adds `updatedAt` and `googleTaskId`.
  - IPC: `createTodo`, `listTodos`, `updateTodo`, `toggleTodo`, `deleteTodo`, `clearCompletedTodos`, `syncTodos`, `onTodosUpdated`.
  - UI: built‑in `todos` and `agenda` panels; Agenda fetches `google_calendar.user.default.calendar.events.list` and merges with due‑dated to‑dos; dashboard seeds Agenda + To‑Do and drops the static Meeting Prep / Up Next / To Do cards.
  - Sync: Google Tasks connector; two‑way sync via executor with last‑write‑wins, full pagination, delete tombstones, single‑flight coalescing, and first‑sync title dedup (keeps multiple same‑title locals distinct by design; completed↔completed matches, active↔completed does not).
  - Agent: `TodosPort` and `manage_todos` tool (list/create/update/toggle/delete/clear_completed/sync); each turn is primed with the sorted open list.

- **Bug Fixes**
  - Sync robustness: paginate fully and abort on partial pages; tolerant insert linking; tombstones for local deletes; coalesce overlaps; reconcile against the live store; skip remote PATCH for mid‑sync deletions; remove orphaned remotes created mid‑sync; tombstone‑gated PATCH skip avoids dropping first‑sync title‑dedup updates.
  - UI/store safety: calendar‑day overdue semantics; guarded fetch ordering and shared push subscription; Agenda keeps events on refresh, handles a null bridge, guards out‑of‑order responses, ticks every 60s, fetches from start of local day, paginates the full window, and flags when results are truncated on very busy calendars; prevent double‑submit; surface real sync errors; show initial load errors; clear stale error on push; failed optimistic edits roll back exactly if no push landed, or reconcile via fetch when superseded.
  - Dedup: title matching excludes active↔completed pairs to avoid silent reopen/auto‑complete; completed↔completed now collapses correctly.
  - Agent: manage_todos “list” now sorts like turn priming; unknown delete returns an error.
  - Logout safety: suppress `onTodosUpdated` after `TodoManager` is closed; renderer logout reset is pre‑warmed and guarded on the current auth phase to avoid clobbering a fast re‑login.

<sup>Written for commit 9aaa6238cb34aff7e75e7a59fd92ed909aae92a9. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/kyh/inteligir/pull/346?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: apps/desktop/src/shared/todo.ts [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+/** Whole-calendar-day delta between two epoch-ms instants (local time), so
+ * "due at 11pm tonight" reads as Today, not Overdue. */
+function dayDelta(from: number, to: number): number {
+  return Math.round((startOfDay(to) - startOfDay(from)) / MS_PER_DAY);
+}
+
+export function isOverdue(todo: Todo, now: number): boolean {
+  return !todo.done && todo.dueAt !== null && todo.dueAt < now;
+}
```
[the remark this answers is not in the document]
Comment (kyh) [reply apps/desktop/src/shared/todo.ts:147]: Good catch — fixed in 667d280. `isOverdue` now compares start-of-day to start-of-day, so a date-only (local-noon) item due today no longer reads as overdue in the afternoon; it's overdue only once its due day has fully passed. Added a `todo-helpers` test covering this.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: apps/desktop/src/renderer/stores/todo-store.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    const bridge = getBridge();
+    if (!bridge) return;
+    set({ loading: true, error: null });
+    void bridge
+      .listTodos()
+      .then((result) => set({ todos: result.todos }))
+      .catch((err: unknown) => {
+        set({ error: `Couldn't load to-dos: ${errorMessage(err)}` });
+      })
+      .finally(() => set({ loading: false }));
```
[the remark this answers is not in the document]
Comment (kyh) [reply apps/desktop/src/renderer/stores/todo-store.ts]: Fixed in 9e650e1. `fetchTodos` now captures a monotonic `snapshotSeq` at request time and only applies its result if no push/mutation bumped the sequence meanwhile — so a slow `listTodos` can't clobber a fresher snapshot. Every push handler and optimistic mutation increments the counter.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: apps/desktop/src/renderer/shell/builtin/todo-panel.tsx [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+  const todos = useTodoStore((s) => s.todos);
+  const init = useTodoStore((s) => s.init);
+  const toggleTodo = useTodoStore((s) => s.toggleTodo);
+  const deleteTodo = useTodoStore((s) => s.deleteTodo);
+  const clearCompleted = useTodoStore((s) => s.clearCompleted);
+  const [creating, setCreating] = useState(false);
+
+  // init fetches the snapshot and subscribes to onTodosUpdated, so items the
+  // agent adds/completes appear without a remount.
+  useEffect(() => init(), [init]);
```
[the remark this answers is not in the document]
Comment (kyh) [reply apps/desktop/src/renderer/shell/builtin/todo-panel.tsx:134]: I believe this is a false positive. Both effects use the concise-arrow form `useEffect(() => init(), [init])`, which **returns** `init()`'s value — and `init()` returns the `onTodosUpdated` unsubscribe. React therefore registers that unsubscribe as the effect's cleanup and calls it on unmount, so the listener is detached (and `init` is a stable zustand action, so the effect runs once per mount). This matches the existing `TaskPanel` pattern. A leak would only occur with a block body (`() => { init(); }`) that drops the return, which isn't the case here. Leaving the code as-is.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: apps/desktop/src/main/todos/google-tasks-sync.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        pushed++;
+      }
+    } catch (err) {
+      console.warn(`[todos] failed to export todo ${create.localId}:`, err);
+    }
+  }
+
+  // 4. Apply every local change in one atomic write (imports + remote-wins
+  // updates + export links; remote-deleted removals).
+  mgr.applySync([...plan.localUpserts, ...links], plan.localDeletes);
```
[the remark this answers is not in the document]
Comment (kyh) [reply apps/desktop/src/main/todos/google-tasks-sync.ts]: Fixed in e875f98. The list is now fully paginated (`listAllRemoteTasks` follows `nextPageToken`), so `planSync` sees the complete remote set and "linked-local missing remotely" genuinely means deleted. Crucially, any page error throws and aborts the sync *before* any local mutation, so a truncated/partial pull can never be mistaken for remote deletions.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: apps/desktop/src/main/todos/google-tasks-sync.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+  return { ok: true, pulled: plan.localUpserts.length, pushed, deleted: plan.localDeletes.length };
+}
+
+async function insertRemote(fields: GoogleTaskWriteFields): Promise<GoogleTask | null> {
+  const created = await widgetCallTool(INSERT_TOOL, { tasklist: DEFAULT_LIST, ...fields });
+  // insert returns the created task object; reuse the list parser on a
+  // single-element array so the same validation applies.
+  return parseGoogleTasks([created])[0] ?? null;
```
[the remark this answers is not in the document]
Comment (kyh) [reply apps/desktop/src/main/todos/google-tasks-sync.ts]: Fixed in e875f98. Added a tolerant `extractInsertedId` that links the local row on *any* successful insert returning an id (bare task object or one wrapped under `result`), with `updated` best-effort. So an insert that succeeds now always links, preventing the duplicate re-export on the next sync. (If Google ever returned no id at all we log and skip — unavoidable, but that shouldn't happen for a created resource.)

---
_Generated by [Claude Code](https://claude.ai/code)_

File: apps/desktop/src/shared/google-tasks.ts [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    }
+  }
+
+  // Remote tasks with no local counterpart — import as new local todos.
+  for (const task of remote) {
+    if (matched.has(task.id)) continue;
+    plan.localUpserts.push(
+      applyRemote(task, { id: newId(), priority: DEFAULT_PRIORITY, createdAt: now }),
+    );
+  }
```
[the remark this answers is not in the document]
Comment (kyh) [reply apps/desktop/src/shared/google-tasks.ts:245]: Fixed in e875f98 with tombstones. When a linked todo is deleted (or cleared via clear-completed), `TodoManager` records its `googleTaskId` in `todo-tombstones.json`. The next sync (1) deletes those remote tasks, (2) excludes tombstoned ids from `planSync`'s input so they aren't re-imported even if the remote delete hasn't landed yet, and (3) clears the tombstone once the remote task is gone. Covered by new manager tests.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: apps/desktop/src/main/todos/google-tasks-sync.ts [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    }
+  }
+
+  // 6. Apply every local change in one atomic write, then drop settled
+  // tombstones.
+  mgr.applySync([...plan.localUpserts, ...links], plan.localDeletes);
+  mgr.clearTombstones(clearedTombstones);
+
+  return { ok: true, pulled: plan.localUpserts.length, pushed, deleted: deletedRemote };
+}
```
[the remark this answers is not in the document]
Comment (kyh) [reply apps/desktop/src/main/todos/google-tasks-sync.ts:224]: Fixed in e830c3c. `syncTodosWithGoogle` now coalesces concurrent callers onto a single in-flight run via a module-level promise guard (the real work moved to `runSync`, cleared in `.finally`). A double-click or a UI sync overlapping the agent's `manage_todos` sync both await the same result instead of each reading the snapshot and double-exporting.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: apps/desktop/src/renderer/stores/todo-store.ts [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+    if (!bridge) return;
+    const before = get().todos;
+    snapshotSeq++;
+    set((s) => ({ todos: s.todos.filter((t) => !t.done) }));
+    try {
+      await bridge.clearCompletedTodos();
+    } catch (err) {
+      set({ todos: before });
+      toast.error(`Couldn't clear completed: ${errorMessage(err)}`);
+    }
```
[the remark this answers is not in the document]
Comment (kyh) [reply apps/desktop/src/renderer/stores/todo-store.ts:180]: Fixed in e830c3c. The update/toggle/delete/clear paths no longer restore a captured `before` snapshot on failure — they call `fetchTodos()` to reconcile from the authoritative main store, which is itself guarded by the snapshot sequence. So a concurrent `onTodosUpdated` push that landed mid-request is preserved rather than clobbered by a stale rollback.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: apps/desktop/src/main/todos/google-tasks-sync.ts [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+      }
+    } catch (err) {
+      console.warn(`[todos] failed to export todo ${create.localId}:`, err);
+    }
+  }
+
+  // 6. Apply every local change in one atomic write, then drop settled
+  // tombstones.
+  mgr.applySync([...plan.localUpserts, ...links], plan.localDeletes);
+  mgr.clearTombstones(clearedTombstones);
```
[the remark this answers is not in the document]
Comment (kyh) [reply apps/desktop/src/main/todos/google-tasks-sync.ts:216]: Fixed in 30bd5a1. `applySync` now reconciles against the **live** store, not the opening snapshot: an export link (or remote-wins update) is only applied to a row that still exists, so a todo deleted mid-sync is no longer resurrected. Additionally, after exporting, the sync re-reads the live store and deletes any just-created remote task whose local row was deleted mid-sync, so it isn't re-imported on the next run. Covered by a new "does not resurrect a row deleted from the current store" test.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: apps/desktop/src/main/todos/google-tasks-sync.ts [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+        console.warn(`[todos] insert for ${create.localId} returned no id; not linked`);
+      }
+    } catch (err) {
+      console.warn(`[todos] failed to export todo ${create.localId}:`, err);
+    }
+  }
+
+  // 6. Apply every local change in one atomic write, then drop settled
+  // tombstones.
+  mgr.applySync([...plan.localUpserts, ...links], plan.localDeletes);
```
[the remark this answers is not in the document]
Comment (kyh) [reply apps/desktop/src/main/todos/google-tasks-sync.ts]: Fixed in 30bd5a1. `applySync` now compares `updatedAt` against the live row: if a concurrent local edit made the live row newer than the plan's remote-wins update/link, the local content is kept and only the `googleTaskId` link is adopted (so the edit survives and the item isn't re-exported). planSync also split its output into `localImports` vs `localUpdates` so applySync can treat brand-new imports and existing-row updates correctly. New "keeps a concurrently-edited newer row but adopts the link" test covers this.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: apps/desktop/src/main/todos/google-tasks-sync.ts [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+      await widgetCallTool(PATCH_TOOL, {
+        tasklist: DEFAULT_LIST,
+        task: update.googleTaskId,
+        ...update.fields,
+      });
+      pushed++;
+    } catch (err) {
+      console.warn(`[todos] failed to push update for ${update.googleTaskId}:`, err);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (kyh) [reply apps/desktop/src/main/todos/google-tasks-sync.ts:166]: Fixed in ddad25a. Before the push loop, the sync now re-reads the live store and skips a `remoteUpdates` PATCH whose `googleTaskId` no longer corresponds to a live linked todo — so a todo deleted mid-sync isn't PATCHed remotely. The delete is tombstoned, so the next sync removes the remote task. This mirrors the orphan re-check the export loop already does.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: apps/desktop/src/renderer/shell/builtin/todo-panel.tsx [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+      {lastSync && !lastSync.ok && (
+        <div className="mb-2 rounded-[10px] bg-hover px-2.5 py-2 text-[10px] text-muted-foreground">
+          Couldn&apos;t sync: {lastSync.error}. If Google Tasks isn&apos;t connected yet, add it in
+          Extensions.
+        </div>
+      )}
+
+      {sorted.length === 0 && !creating && (
+        <div className="py-4 text-center text-[10px] text-muted-foreground">Nothing to do yet</div>
+      )}
```
[the remark this answers is not in the document]
Comment (kyh) [reply apps/desktop/src/renderer/shell/builtin/todo-panel.tsx:181]: Fixed in ddad25a. `TodoPanel` now reads the store's `error` and renders it in place of the "Nothing to do yet" empty state, so a failed initial load shows the load error instead of looking like an empty list.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: apps/desktop/src/renderer/stores/todo-store.ts [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+      // superseded it.
+      .then((result) =>
+        token === fetchToken && seq === snapshotSeq ? set({ todos: result.todos }) : undefined,
+      )
+      .catch((err: unknown) => {
+        if (token === fetchToken) set({ error: `Couldn't load to-dos: ${errorMessage(err)}` });
+      })
+      .finally(() => {
+        if (token === fetchToken) set({ loading: false });
+      });
```
[the remark this answers is not in the document]
Comment (kyh) [reply apps/desktop/src/renderer/stores/todo-store.ts:102]: Fixed in 7e26d9f. The `onTodosUpdated` push handler now clears `error` when it applies a fresh snapshot, so a stale "Couldn't load to-dos" message can't survive a successful refresh.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: apps/desktop/src/renderer/stores/todo-store.ts [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+      get().fetchTodos();
+    }
+    return () => {
+      subscriberCount--;
+      if (subscriberCount === 0) {
+        sharedUnsubscribe?.();
+        sharedUnsubscribe = null;
+      }
+    };
+  },
```
[the remark this answers is not in the document]
Comment (kyh) [reply apps/desktop/src/renderer/stores/todo-store.ts:82]: Fixed in 7e26d9f. Added a `reset()` action to the todo-store (clears the snapshot and invalidates any in-flight fetch/push) and call it from the agent-store's `logged_out` handler — alongside the existing voice-store reset — so the next user never sees the prior user's items. It's a lazy `import()` to keep todo-store's UI deps out of agent-store's static graph.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: apps/desktop/src/agent/todos/extension.ts [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+      switch (p.action) {
+        case "list": {
+          const all = todos.getTodos();
+          if (all.length === 0) return textResult("The to-do list is empty.");
+          const lines = all.map((t) => {
+            const box = t.done ? "[x]" : "[ ]";
+            const due = t.dueAt !== null ? ` · due ${formatDue(t.dueAt, now)}` : "";
+            return `${box} ${t.title} (${t.id}) — ${t.priority}${due}`;
+          });
+          return textResult(lines.join("\n"));
```
[the remark this answers is not in the document]
Comment (kyh) [reply apps/desktop/src/agent/todos/extension.ts:74]: Fixed in 7e26d9f. The `manage_todos` `list` action now runs `sortTodos` too, so it agrees with the turn-priming order (overdue/soonest-due first).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: apps/desktop/src/main/todos/google-tasks-sync.ts [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+  // removes the remote task. (Mirrors the orphan re-check the export loop does.)
+  const liveLinkedIds = new Set(
+    mgr
+      .getTodos()
+      .map((t) => t.googleTaskId)
+      .filter((id): id is string => id !== null),
+  );
+  let pushed = 0;
+  for (const update of plan.remoteUpdates) {
+    if (!liveLinkedIds.has(update.googleTaskId)) continue;
```
[the remark this answers is not in the document]
Comment (kyh) [reply apps/desktop/src/main/todos/google-tasks-sync.ts]: Good catch — this was a regression my own mid-sync-delete guard introduced. Fixed in 304fd54: the guard now skips a PATCH only when the `googleTaskId` is freshly **tombstoned** (the real "deleted mid-sync" signal), rather than requiring it to already be on a linked live todo. A first-sync title-dedup PATCH targets a remote whose local row isn't linked until the `applySync` later in the same run, so the previous link-presence check wrongly dropped it; the tombstone check lets it through while still skipping PATCHes for todos the user actually deleted.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: apps/desktop/src/renderer/shell/builtin/agenda-panel.tsx [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+        setCalError(res.error);
+      }
+      setLoaded(true);
+    };
+    void bridge
+      .widgetCallTool({
+        tool: CALENDAR_TOOL,
+        input: {
+          calendarId: "primary",
+          timeMin: start.toISOString(),
```
[the remark this answers is not in the document]
Comment (kyh) [reply apps/desktop/src/renderer/shell/builtin/agenda-panel.tsx]: Fixed in 304fd54. The agenda calendar fetch now sets `timeMin` to the start of the local day instead of the current instant, so meetings earlier today load and the "Today" group is complete.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: apps/desktop/src/renderer/shell/builtin/agenda-panel.tsx [hunk 18]
```diff
@@ -0,0 +1,210 @@
...
+          timeMax: timeMax.toISOString(),
+          singleEvents: true,
+          orderBy: "startTime",
+          maxResults: 50,
+        },
```
[the remark this answers is not in the document]
Comment (kyh) [reply apps/desktop/src/renderer/shell/builtin/agenda-panel.tsx]: Fixed in 977d973. The agenda calendar fetch now paginates the full 14-day window (follows `nextPageToken`, 250/page, 20-page safety cap) instead of a single 50-event page, so a busy window isn't silently truncated.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: apps/desktop/src/renderer/stores/todo-store.ts [hunk 19]
```diff
@@ -0,0 +1,214 @@
...
+    const bridge = getBridge();
+    if (!bridge) return;
+    snapshotSeq++;
+    set((s) => ({ todos: s.todos.filter((t) => !t.done) }));
+    try {
+      await bridge.clearCompletedTodos();
+    } catch (err) {
+      get().fetchTodos();
+      toast.error(`Couldn't clear completed: ${errorMessage(err)}`);
+    }
```
[the remark this answers is not in the document]
Comment (kyh) [reply apps/desktop/src/renderer/stores/todo-store.ts:180]: Fixed in 977d973. This balances against the earlier "rollback overwrites newer pushes" finding: on failure, recovery now rolls back to the exact prior list when `snapshotSeq` is unchanged (no push landed since the optimistic change) — immediate and correct — and only falls back to `fetchTodos()` when a concurrent push superseded it (so the newer state isn't clobbered). Either way the UI no longer stays stuck on a failed edit.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: apps/desktop/src/main/todos/todo-manager.ts [hunk 20]
```diff
@@ -0,0 +1,314 @@
...
+        } else {
+          // Live row is newer (edited mid-sync): keep its content, adopt the link.
+          next.push({ ...current, googleTaskId: update.googleTaskId ?? current.googleTaskId });
+        }
+      }
+      const seen = new Set(next.map((t) => t.id));
+      for (const todo of imports) if (!seen.has(todo.id)) next.push(todo);
+      return next;
+    });
+    this.notifyTodosChanged();
```
[the remark this answers is not in the document]
Comment (kyh) [reply apps/desktop/src/main/todos/todo-manager.ts:283]: Valid — fixed in 2cc0395. Added a one-way `closed` flag to `TodoManager` (mirroring `JsonStore`'s kill switch): `close()` sets it and `notifyTodosChanged()` short-circuits when set, so an in-flight `syncTodosWithGoogle()` holding the pre-logout instance can no longer broadcast the prior user's cached todos over `onTodosUpdated` after teardown. Covered by a new test asserting no broadcast fires after `close()`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: apps/desktop/src/main/todos/google-tasks-sync.ts [hunk 21]
```diff
@@ -0,0 +1,224 @@
...
+  const remoteForPlan = remote.filter((t) => !tombSet.has(t.id));
+  const plan = planSync(local, remoteForPlan, Date.now(), () => crypto.randomUUID());
+
+  // 4. Push updates to already-linked tasks. Per-task failures are tolerated:
+  // the unsynced edit stays "newer" locally, so the next run retries it.
+  // Re-read tombstones first: if the user deleted a linked todo while we were
+  // pulling, its googleTaskId is freshly tombstoned — skip that PATCH (the next
+  // sync deletes the remote task). We key on tombstones, NOT link-presence, so
+  // a first-sync title-dedup PATCH (whose local row isn't linked until the
+  // applySync below) still goes through.
```
[the remark this answers is not in the document]
Comment (kyh) [reply apps/desktop/src/main/todos/google-tasks-sync.ts:151]: Leaving this one as-is by design — flagging the reasoning rather than adding a rollback layer.

Two-system reconciliation can't be made atomic without distributed transactions, which the Google Tasks API doesn't offer. The remote→local ordering is inherent: linking a local row requires the id the `insert` returns, so the remote write must come first. What we *can* control is the blast radius of a torn write, and the design already bounds it:

- `applySync`/`clearTombstones` are synchronous in-memory mutations over a `JsonStore` (disk is a debounced, non-throwing follow-up). They don't fail mid-flight under normal operation, and if one did throw, `runSync` rejects → `syncTodosWithGoogle().catch` → `{ ok: false, error }`, so the failure is surfaced, not swallowed.
- The divergent state self-heals on the next sync rather than compounding:
  - An unlinked local whose remote task was created but whose link wasn't persisted is **not** re-duplicated — first-sync title dedup (hardened in 9aaa623) adopts the existing same-title remote instead of creating a second one.
  - Stale tombstones don't accumulate failed DELETEs — a tombstone is cleared the moment its remote task is already absent (`!remoteIds.has(...)`), and the DELETE path tolerates 404.

So a partial write converges on retry instead of drifting. A real rollback/2PC layer would add significant complexity for a window that's both near-impossible (local sync write) and already self-correcting. Happy to revisit if you'd rather persist links incrementally as each insert lands — just didn't want to add that machinery speculatively.

---
_Generated by [Claude Code](https://claude.ai/code)_
